content/3_Silver/Sorting_Custom.mdx

```diff
@@ -640,6 +639,198 @@ public class Sol {
 
 </LanguageSection>
 
+## Coordinate Compression
+
+Coordinate compression describes the process of replacing each value in a list with its index if that
+list was sorted. For example, the list $\{7, 3, 4, 1\}$ would be compressed to
+$\{3, 1, 2, 0\}$. Notice that $1$ is the least value in the first list, so it
+becomes $0$, and $7$ is the greatest value, so it becomes $3$, the largest index
+in the list.
+
+When we have values that can take on a large range of possible values, but we
+only care about their relative order (for example, if we have to know if one
+value is above another), coordinate compression is a simple way to help a lot
+with implementation. If we have a set of integers ranging from $0$ to $10^9$, we
+can't use them as array indices because we'd have to create an array of size
+$10^9$, which would surely cause a `Memory Limit Exceeded` verdict. However, if
+there are only $N \leq 10^6$ such integers, we can coordinate-compress their
+values, which guarantees that the values will all be in the range from $0$ to
+$N-1$, which _can_ be used as array indices.
+
+<FocusProblem problem="sample2" />
+
+A good example of coordinate compression in action is in the solution of _USACO
+Rectangular Pasture._ Again, we won't delve into the full solution but instead
+discuss how coordinate compression is applied. Since the solution uses
+[2D-prefix](/silver/prefix-sums-2) sums (another Silver topic), it is helpful if
+all point coordinates are coordinate-compressed to the range $0$ to $N-1$ so
+they can be used as array indices. Without coordinate compression, creating a large enough array would result in a `Memory Limit Exceeded` verdict.
+
+Below you will find the solution to Rectangular Pasture, which uses
+coordinate compression at the beginning. Observe how a custom comparator is used to sort the
+points:
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+BeginCodeSnip{}
+#include <iostream>
+#include <algorithm>
+using namespace std;
+ 
+typedef pair<int,int> Point;
+bool ycomp(Point p, Point q) { return p.second < q.second; }
+ 
+const int MAX_N = 2500;
+int N, Psum[MAX_N+1][MAX_N+1];
+Point P[MAX_N];
+ 
+int rsum(int x1, int y1, int x2, int y2)
+{
+  return Psum[x2+1][y2+1] - Psum[x2+1][y1] - Psum[x1][y2+1] + Psum[x1][y1];
+}
+ 
+int main(void)
+{
+  cin >> N;
+  for (int i=0; i<N; i++) {
+    int x, y;
+    cin >> x >> y;
+    P[i] = make_pair(x,y);
+  }
+EndCodeSnip
+  sort(P, P+N); 
+  for (int i=0; i<N; i++) P[i].first = i+1;
+  sort(P, P+N, ycomp); 
```

Comment:
Also up to you: `sort(P, P+N, [](Point a, Point b){
           return a.second < b.second;
}
);`

Reply:
Fine as is.